

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Guide / Tutorial &mdash; BRKGA-MP-IPR 0.9 documentation</title>
  

  
  
    <link rel="shortcut icon" href="_static/favicon.ico"/>
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
        <script src="_static/add_target.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/custom_theme.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="BRKGA-MP-IPR License" href="page_license.html" />
    <link rel="prev" title="BRKGA-MP-IPR Guide and Documentation - Python version" href="index.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> BRKGA-MP-IPR
          

          
            
            <img src="_static/logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Guide / Tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#installation-and-tests">Installation and tests</a></li>
<li class="toctree-l2"><a class="reference internal" href="#tl-dr">TL;DR</a></li>
<li class="toctree-l2"><a class="reference internal" href="#getting-started">Getting started</a></li>
<li class="toctree-l2"><a class="reference internal" href="#first-things-first-the-decoder-function">First things first: the decoder function</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#the-chromosome-or-vector-of-doubles">The chromosome or vector of doubles</a></li>
<li class="toctree-l3"><a class="reference internal" href="#back-to-the-decoder">Back to the decoder</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#building-brkga-mp-ipr-algorithm-object">Building BRKGA-MP-IPR algorithm object</a></li>
<li class="toctree-l2"><a class="reference internal" href="#initialization-and-warm-start-solutions">Initialization and Warm-start solutions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#warm-start-solutions">Warm-start solutions</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#optimization-time-evolving-the-population">Optimization time: evolving the population</a></li>
<li class="toctree-l2"><a class="reference internal" href="#accessing-solutions-chromosomes">Accessing solutions/chromosomes</a></li>
<li class="toctree-l2"><a class="reference internal" href="#implicit-path-relink">Implicit Path Relink</a></li>
<li class="toctree-l2"><a class="reference internal" href="#shaking-and-resetting">Shaking and Resetting</a></li>
<li class="toctree-l2"><a class="reference internal" href="#multi-population-and-migration">Multi-population and migration</a></li>
<li class="toctree-l2"><a class="reference internal" href="#simulating-the-standard-brkga">Simulating the standard BRKGA</a></li>
<li class="toctree-l2"><a class="reference internal" href="#reading-and-writing-parameters">Reading and writing parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="#probable-valuable-tips">(Probable Valuable) Tips</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#algorithm-warmup">Algorithm warmup</a></li>
<li class="toctree-l3"><a class="reference internal" href="#complex-decoders-and-timing">Complex decoders and timing</a></li>
<li class="toctree-l3"><a class="reference internal" href="#multi-threading">Multi-threading</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="page_license.html">BRKGA-MP-IPR License</a></li>
<li class="toctree-l1"><a class="reference internal" href="page_todo.html">Todo List</a></li>
<li class="toctree-l1"><a class="reference internal" href="page_md_src_contributing.html">Contributing</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="api.html">API</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">BRKGA-MP-IPR</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Guide / Tutorial</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/page_guide.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="guide-tutorial">
<span id="doxid-guide"></span><span id="index-0"></span><h1>Guide / Tutorial<a class="headerlink" href="#guide-tutorial" title="Permalink to this headline">¶</a></h1>
<div class="section" id="installation-and-tests">
<span id="doxid-guide-1guide-installation"></span><h2>Installation and tests<a class="headerlink" href="#installation-and-tests" title="Permalink to this headline">¶</a></h2>
<p>BRKGA-MP-IPR was developed using &gt;= Python 3.7.2, especially using the new
<code class="docutils literal notranslate"><span class="pre">enum</span></code> capabilities. The parameters’ loading and writing functions may fail
on Python 3.6 or previous versions. However, the main algorithm functions
work fine on Python 3.6, by providing BrkgaParams manually (or implementing
your own parameter loading).</p>
<p>Assuming you have the correct Python version, the installation is pretty
straightforward using Pypi:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ pip3.7 search brkga
brkga-mp-ipr <span class="o">(</span><span class="m">0</span>.9<span class="o">)</span>  - The Multi-Parent Biased Random-Key Genetic Algorithm with Implict Path Relink

$ pip3.7 install brkga-mp-ipr
Collecting brkga-mp-ipr
...
Installing collected packages: brkga-mp-ipr
Successfully installed brkga-mp-ipr-0.9

$ python3.7
Python <span class="m">3</span>.7.5 <span class="o">(</span>default, Oct <span class="m">19</span> <span class="m">2019</span>, <span class="m">01</span>:20:12<span class="o">)</span>
Type <span class="s2">&quot;help&quot;</span>, <span class="s2">&quot;copyright&quot;</span>, <span class="s2">&quot;credits&quot;</span> or <span class="s2">&quot;license&quot;</span> <span class="k">for</span> more information.
&gt;&gt;&gt; from brkga_mp_ipr.types_io import load_configuration
&gt;&gt;&gt; from brkga_mp_ipr.algorithm import BrkgaMpIpr
&gt;&gt;&gt; BrkgaMpIpr
&lt;class <span class="s1">&#39;brkga_mp_ipr.algorithm.BrkgaMpIpr&#39;</span>&gt;
&gt;&gt;&gt; load_configuration
&lt;<span class="k">function</span> load_configuration at 0x10620e320&gt;
&gt;&gt;&gt; help<span class="o">(</span>load_configuration<span class="o">)</span>
Help on <span class="k">function</span> load_configuration in module brkga_mp_ipr.types_io:

load_configuration<span class="o">(</span>configuration_file: str<span class="o">)</span> -&gt; <span class="o">(</span>&lt;class <span class="s1">&#39;brkga_mp_ipr.types.BrkgaParams&#39;</span>&gt;, &lt;class <span class="s1">&#39;brkga_mp_ipr.types.ExternalControlParams&#39;</span>&gt;<span class="o">)</span>
    Loads the parameters from <span class="sb">`</span>configuration_file<span class="sb">`</span> returning them as a tuple.

    Args:
        configuration_file <span class="o">(</span>str<span class="o">)</span>: plain text file containing the configuration.

    Returns:
        A tuple containing a <span class="sb">`</span>BrkgaParams<span class="sb">`</span> and a <span class="sb">`</span>ExternalControlParams<span class="sb">`</span> object.

    Raises:
        IsADirectoryError: If <span class="sb">`</span>configuration_file<span class="sb">`</span> is a folder.

        FileNotFoundError: If <span class="sb">`</span>configuration_file<span class="sb">`</span> does not exist.

        LoadError: In cases of missing data or bad-formatted data.
</pre></div>
</div>
<p>BRKGA-MP-IPR also provides a thorough unit testing that aims to harden and make
the code ready for production environments.
You can use builtin
<a class="reference external" href="https://docs.python.org/3.7/library/unittest.html">unittest</a>,
or yet <a class="reference external" href="https://www.pytest.org">pytest</a>
or <a class="reference external" href="https://tox.readthedocs.io">Tox</a>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The tests take about 10 minutes, mainly because the permutation path relink.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>It is a hard test to test algorithms that use random signals. In
BRKGA-MP-IPR, the tests are carefully designed to ensure repeatability. For
that, we use the Mersenne Twister
<a class="reference external" href="https://en.wikipedia.org/wiki/Mersenne_Twister">[1]</a>
<a class="reference external" href="http://dx.doi.org/10.1145/272991.272995">[2]</a> as our standard random
generator number engine, particularly the <a class="reference external" href="https://docs.python.org/3.7/library/random.html">version that comes with
Python 3.7</a>.
However, it may happen that such engine has slightly different
implementations across platforms and, therefore, the tests may fail. The
current version was tested on 64-bit platforms (Mac OS X, GNU/Linux, and
Windows 10).</p>
</div>
</div>
<div class="section" id="tl-dr">
<span id="doxid-guide-1guide-tldr"></span><h2>TL;DR<a class="headerlink" href="#tl-dr" title="Permalink to this headline">¶</a></h2>
<p>The best way to keep it short is to look in the on examples on <a class="reference external" href="https://github.com/ceandrade/brkga_mp_ipr_python/tree/master/examples">the git repo.</a>
Let’s take a look into
<a class="reference external" href="https://github.com/ceandrade/brkga_mp_ipr_python/blob/master/examples/tsp/main_minimal.py">main_minimal.cpp</a>,
which solves the
<a class="reference external" href="https://en.wikipedia.org/wiki/Travelling_salesman_problem">Traveling Salesman Problem (TSP)</a>.
This is a trimmed copy:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">sys</span>

<span class="kn">from</span> <span class="nn">brkga_mp_ipr.enums</span> <span class="kn">import</span> <span class="n">Sense</span>
<span class="kn">from</span> <span class="nn">brkga_mp_ipr.types_io</span> <span class="kn">import</span> <span class="n">load_configuration</span>
<span class="kn">from</span> <span class="nn">brkga_mp_ipr.algorithm</span> <span class="kn">import</span> <span class="n">BrkgaMpIpr</span>

<span class="kn">from</span> <span class="nn">tsp_instance</span> <span class="kn">import</span> <span class="n">TSPInstance</span>
<span class="kn">from</span> <span class="nn">tsp_decoder</span> <span class="kn">import</span> <span class="n">TSPDecoder</span>

<span class="k">def</span> <span class="nf">main</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">4</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Usage: python main_minimal.py &lt;seed&gt; &lt;config-file&gt; &quot;</span>
            <span class="s2">&quot;&lt;num-generations&gt; &lt;tsp-instance-file&gt;&quot;</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

    <span class="n">seed</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
    <span class="n">configuration_file</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
    <span class="n">num_generations</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
    <span class="n">instance_file</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span>

    <span class="n">instance</span> <span class="o">=</span> <span class="n">TSPInstance</span><span class="p">(</span><span class="n">instance_file</span><span class="p">)</span>

    <span class="n">decoder</span> <span class="o">=</span> <span class="n">TSPDecoder</span><span class="p">(</span><span class="n">instance</span><span class="p">)</span>

    <span class="n">brkga_params</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">load_configuration</span><span class="p">(</span><span class="n">configuration_file</span><span class="p">)</span>

    <span class="n">brkga</span> <span class="o">=</span> <span class="n">BrkgaMpIpr</span><span class="p">(</span>
        <span class="n">decoder</span><span class="o">=</span><span class="n">decoder</span><span class="p">,</span>
        <span class="n">sense</span><span class="o">=</span><span class="n">Sense</span><span class="o">.</span><span class="n">MINIMIZE</span><span class="p">,</span>
        <span class="n">seed</span><span class="o">=</span><span class="n">seed</span><span class="p">,</span>
        <span class="n">chromosome_size</span><span class="o">=</span><span class="n">instance</span><span class="o">.</span><span class="n">num_nodes</span><span class="p">,</span>
        <span class="n">params</span><span class="o">=</span><span class="n">brkga_params</span>
    <span class="p">)</span>

    <span class="n">brkga</span><span class="o">.</span><span class="n">initialize</span><span class="p">()</span>

    <span class="n">brkga</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span><span class="n">num_generations</span><span class="p">)</span>

    <span class="n">best_cost</span> <span class="o">=</span> <span class="n">brkga</span><span class="o">.</span><span class="n">get_best_fitness</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Best cost: </span><span class="si">{</span><span class="n">best_cost</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">main</span><span class="p">()</span>
</pre></div>
</div>
<p>You can identify the following basic steps:</p>
<ol class="arabic">
<li><p>Create a data structure to hold your input data which is passed to the
decoder function (example
<a class="reference external" href="https://github.com/ceandrade/brkga_mp_ipr_python/blob/master/examples/tsp/tsp_instance.py">tsp_instance.py</a>).
Note that you may not implement a data/instance class but load all needed
information directly on your decoder;</p></li>
<li><p>Create a decoder class. The <code class="docutils literal notranslate"><span class="pre">decode()</span></code> method from this class
translates a chromosome (array of numbers in the interval [0, 1]) to a
solution for your problem. The decoder must return the solution value or cost
to be used as fitness by BRKGA (example
<a class="reference external" href="https://github.com/ceandrade/brkga_mp_ipr_python/blob/master/examples/tsp/tsp_decoder.py">tsp_decoder.py</a>).
Note that the <code class="docutils literal notranslate"><span class="pre">decode()</span></code> method must have the following signature:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">decode</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chromosome</span><span class="p">:</span> <span class="n">BaseChromosome</span><span class="p">,</span> <span class="n">rewrite</span><span class="p">:</span> <span class="nb">bool</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span>
</pre></div>
</div>
<p>where <code class="docutils literal notranslate"><span class="pre">BaseChromosome</span></code> is a class inhereted from <code class="docutils literal notranslate"><span class="pre">list</span></code>. In other words,
you can tread <code class="docutils literal notranslate"><span class="pre">chromosome</span></code> as a simple list of floats;</p>
</li>
<li><p>Load the instance and other relevant data, and instantiate the decoder;</p></li>
<li><p>Read the algorithm parameters using <code class="docutils literal notranslate"><span class="pre">load_configuration()</span></code> or create
a BrkgaParams object by hand;</p></li>
<li><p>Create a <code class="docutils literal notranslate"><span class="pre">BrkgaMpIpr</span></code> algorithm object;</p></li>
<li><p>Call <code class="docutils literal notranslate"><span class="pre">initialize()</span></code> to init the BRKGA state;</p></li>
<li><p>Call <code class="docutils literal notranslate"><span class="pre">evolve()</span></code> to optimize;</p></li>
<li><p>Call <code class="docutils literal notranslate"><span class="pre">get_best_fitness()</span></code> and/or <code class="docutils literal notranslate"><span class="pre">get_best_chromosome()</span></code> to
retrieve the best solution.</p></li>
</ol>
<p><a class="reference external" href="https://github.com/ceandrade/brkga_mp_ipr_python/blob/master/examples/tsp/main_minimal.py">main_minimal.py</a>
provides a very minimal example to understand the necessary steps to use the
BRKGA-MP-IPR framework. However,
<a class="reference external" href="https://github.com/ceandrade/brkga_mp_ipr_python/blob/master/examples/tsp/main_complete.py">main_complete.py</a>
provides a full-featured code, handy for scientific use, such as
experimentation and paper writing. This code allows fine-grained control of
the optimization, shows several features of BRKGA-MP-IPR such as the resets,
chromosome injection, and others. It also logs
all optimization steps, <em>creating outputs easy to be parsed.</em> <strong>You should use
this code for serious business and experimentation.</strong></p>
</div>
<div class="section" id="getting-started">
<span id="doxid-guide-1guide-getting-started"></span><h2>Getting started<a class="headerlink" href="#getting-started" title="Permalink to this headline">¶</a></h2>
<p>BRKGA-MP-IPR is pretty simple, and you must provide one required <strong>decoder</strong>
object to translate chromosomes to solutions.
In general, such decoder uses the problem information to map a vector of real
numbers in the interval [0,1] to a (valid) solution. In some cases, even
though a valid solution cannot be found, library users apply penalization
factors and push the BRKGA to find valid solutions.</p>
<p>Before you go further, please take a look at the <code class="docutils literal notranslate"><span class="pre">examples</span></code> folder in <a class="reference external" href="https://github.com/ceandrade/brkga_mp_ipr_python">the
git repo</a>.
In this guide, we solve the classical <a class="reference external" href="https://en.wikipedia.org/wiki/Travelling_salesman_problem">Traveling Salesman Problem</a>. Given a set of
cities and the distances between them (full weighted undirect graph), one
must find a minimum-cost tour among all cities, such that each city is
visited only once (i.e., find a Hamiltonian cycle of minimum cost). The
folder has the following structure:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://github.com/ceandrade/brkga_mp_ipr_python/blob/master/examples/tsp/tsp_instance.py">tsp_instance.py</a>:
contains the input data structures and helper functions;</p></li>
<li><p><a class="reference external" href="https://github.com/ceandrade/brkga_mp_ipr_python/blob/master/examples/tsp/tsp_decoder.py">tsp_decoder.py</a>:
contains the decoder function for TSP;</p></li>
<li><p><a class="reference external" href="https://github.com/ceandrade/brkga_mp_ipr_python/blob/master/examples/tsp/greedy_tour.py">greedy_tour.py</a>:
simple heuristic that computes a greedy tour;</p></li>
<li><p><a class="reference external" href="https://github.com/ceandrade/brkga_mp_ipr_python/blob/master/examples/tsp/config.conf">config.conf</a>:
example of parameter settings;</p></li>
<li><p><a class="reference external" href="https://github.com/ceandrade/brkga_mp_ipr_python/blob/master/examples/tsp/main_minimal.py">main_minimal.py</a>:
minimal code useful to understand and test the framework.
<strong>You should start here!</strong> Please take a look on this file before continue
this tutorial;</p></li>
<li><p><a class="reference external" href="https://github.com/ceandrade/brkga_mp_ipr_python/blob/master/examples/tsp/main_complete.py">main_complete.py</a>:
full-featured code, handy for scientific use, such as
experimentation and paper writing. This code allows fine-grained control of
the optimization, shows several features of BRKGA-MP-IPR such as the
path-relinking calls, resets, chromosome injection, and others. It also logs
all optimization steps, _creating outputs easy to be parsed._
<strong>You should use this code for serious business and experimentation;</strong></p></li>
<li><p><a class="reference external" href="https://github.com/ceandrade/brkga_mp_ipr_python/tree/master/examples/tsp/instances">instances</a>:
folder containing some TSP instances for testing.</p></li>
</ul>
<p>When you call
<a class="reference external" href="https://github.com/ceandrade/brkga_mp_ipr_python/blob/master/examples/tsp/main_minimal.py">main_minimal.py</a>
or
<a class="reference external" href="https://github.com/ceandrade/brkga_mp_ipr_python/blob/master/examples/tsp/main_complete.py">main_complete.py</a>:
without arguments, they show the usage. For example, assuming you are using
a terminal:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ python3.7 main_minimal.py
Usage: python main_minimal.py &lt;seed&gt; &lt;config-file&gt; &lt;num-generations&gt; &lt;tsp-instance-file&gt;

$ python3.7 main_complete.py
Usage:
  main_complete.py -c &lt;config_file&gt; -s &lt;seed&gt; -r &lt;stop_rule&gt; -a &lt;stop_arg&gt; -t &lt;max_time&gt; -i &lt;instance_file&gt; <span class="o">[</span>--no_evolution<span class="o">]</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><a class="reference external" href="https://github.com/ceandrade/brkga_mp_ipr_python/blob/master/examples/tsp/main_complete.py">main_complete.py</a>
uses the <a class="reference external" href="https://github.com/docopt/docopt">DocOpt package</a>.
Please, install it before run this script.</p>
</div>
<p>So, this is a possible output whe calling <code class="docutils literal notranslate"><span class="pre">main_minimal</span></code> :</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ python3.7 main_minimal.py <span class="m">27000001</span> config.conf <span class="m">100</span> instances/brazil58.dat
Reading data...
Reading parameters...
Building BRKGA data and initializing...
Evolving <span class="m">100</span> generations...
Best cost: <span class="m">89375</span>.0
</pre></div>
</div>
<p>For <code class="docutils literal notranslate"><span class="pre">main_complete</span></code>, the output is more verbose, since we want to capture
as much information as possible to do some statistical analysis. The output
should be something close to this:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ python3.7 main_complete.py -c config.conf -s <span class="m">2700001</span> -r Generations -a <span class="m">100</span> -t <span class="m">60</span> -i instances/brazil58.dat
------------------------------------------------------
&gt; Experiment started at <span class="m">2019</span>-11-21 <span class="m">18</span>:37:03.023320
&gt; Instance: instances/brazil58.dat
&gt; Configuration: config.conf
&gt; Algorithm Parameters:
&gt;  -population_size <span class="m">2000</span>
&gt;  -elite_percentage <span class="m">0</span>.3
&gt;  -mutants_percentage <span class="m">0</span>.15
&gt;  -num_elite_parents <span class="m">2</span>
&gt;  -total_parents <span class="m">3</span>
&gt;  -bias_type LOGINVERSE
&gt;  -num_independent_populations <span class="m">3</span>
&gt;  -pr_number_pairs <span class="m">0</span>
&gt;  -pr_minimum_distance <span class="m">0</span>.15
&gt;  -pr_type PERMUTATION
&gt;  -pr_selection BESTSOLUTION
&gt;  -alpha_block_size <span class="m">1</span>.0
&gt;  -pr_percentage <span class="m">1</span>.0
&gt;  -exchange_interval <span class="m">200</span>
&gt;  -num_exchange_indivuduals <span class="m">2</span>
&gt;  -reset_interval <span class="m">600</span>
&gt; Seed: <span class="m">2700001</span>
&gt; Stop rule: GENERATIONS
&gt; Stop argument: <span class="m">100</span>
&gt; Maximum <span class="nb">time</span> <span class="o">(</span>s<span class="o">)</span>: <span class="m">60</span>.0
------------------------------------------------------

<span class="o">[</span><span class="m">2019</span>-11-21 <span class="m">18</span>:37:03.023403<span class="o">]</span> Reading TSP data...
Number of nodes: <span class="m">58</span>

<span class="o">[</span><span class="m">2019</span>-11-21 <span class="m">18</span>:37:03.024271<span class="o">]</span> Generating initial tour...
Initial cost: <span class="m">30774</span>.0

<span class="o">[</span><span class="m">2019</span>-11-21 <span class="m">18</span>:37:03.025056<span class="o">]</span> Building BRKGA data...
New population size: <span class="m">580</span>

<span class="o">[</span><span class="m">2019</span>-11-21 <span class="m">18</span>:37:03.025311<span class="o">]</span> Initializing BRKGA data...

<span class="o">[</span><span class="m">2019</span>-11-21 <span class="m">18</span>:37:03.193528<span class="o">]</span> Warming up...

<span class="o">[</span><span class="m">2019</span>-11-21 <span class="m">18</span>:37:04.117741<span class="o">]</span> Evolving...
* Iteration <span class="p">|</span> Cost <span class="p">|</span> CurrentTime
* <span class="m">1</span> <span class="p">|</span> <span class="m">30774</span> <span class="p">|</span> <span class="m">0</span>.40
* <span class="m">74</span> <span class="p">|</span> <span class="m">30759</span> <span class="p">|</span> <span class="m">30</span>.15
* <span class="m">78</span> <span class="p">|</span> <span class="m">30721</span> <span class="p">|</span> <span class="m">31</span>.77
* <span class="m">79</span> <span class="p">|</span> <span class="m">30371</span> <span class="p">|</span> <span class="m">32</span>.18
* <span class="m">81</span> <span class="p">|</span> <span class="m">30350</span> <span class="p">|</span> <span class="m">32</span>.99
<span class="o">[</span><span class="m">2019</span>-11-21 <span class="m">18</span>:37:44.924912<span class="o">]</span> End of optimization

Total number of iterations: <span class="m">100</span>
Last update iteration: <span class="m">81</span>
Total optimization time: <span class="m">40</span>.81
Last update time: <span class="m">32</span>.99
Large number of iterations between improvements: <span class="m">73</span>

% Best tour cost: <span class="m">30350</span>.00
% Best tour: <span class="m">41</span> <span class="m">0</span> <span class="m">29</span> <span class="m">12</span> <span class="m">39</span> <span class="m">24</span> <span class="m">8</span> <span class="m">31</span> <span class="m">19</span> <span class="m">52</span> <span class="m">49</span> <span class="m">3</span> <span class="m">17</span> <span class="m">43</span> <span class="m">23</span> <span class="m">57</span> <span class="m">4</span> <span class="m">26</span> <span class="m">42</span> <span class="m">11</span> <span class="m">56</span> <span class="m">22</span> <span class="m">53</span> <span class="m">54</span> <span class="m">1</span> <span class="m">40</span> <span class="m">34</span> <span class="m">9</span> <span class="m">51</span> <span class="m">50</span> <span class="m">46</span> <span class="m">48</span> <span class="m">2</span> <span class="m">47</span> <span class="m">28</span> <span class="m">35</span> <span class="m">16</span> <span class="m">25</span> <span class="m">18</span> <span class="m">5</span> <span class="m">27</span> <span class="m">32</span> <span class="m">13</span> <span class="m">36</span> <span class="m">33</span> <span class="m">45</span> <span class="m">55</span> <span class="m">44</span> <span class="m">14</span> <span class="m">20</span> <span class="m">38</span> <span class="m">10</span> <span class="m">15</span> <span class="m">21</span> <span class="m">7</span> <span class="m">37</span> <span class="m">30</span> <span class="m">6</span>

Instance,Seed,NumNodes,TotalIterations,TotalTime,LargeOffset,LastUpdateIteration,LastUpdateTime,Cost
brazil58.dat,2700001,58,100,40.81,73,81,32.99,30350
</pre></div>
</div>
<p>I hope by now you got your system set up and running. Let’s see the essential
details on how to use the BRKGA-MP-IPR.</p>
</div>
<div class="section" id="first-things-first-the-decoder-function">
<span id="doxid-guide-1guide-decoder"></span><h2>First things first: the decoder function<a class="headerlink" href="#first-things-first-the-decoder-function" title="Permalink to this headline">¶</a></h2>
<p>The core of the BRKGA algorithm is the definition of a decoder
function/object. The decoder maps the chromosomes (vectors of real numbers in
the interval [0, 1]) to solutions of the problem. In some sense, a decoder is
similar to a <a class="reference external" href="https://en.wikipedia.org/wiki/Kernel_method">kernel function from Support Vector Machines</a> : both functions are used to
project solutions/distances in different spaces.</p>
<p>Here, we have a small difference between the Python/C++ and the Julia
implementations. In the Julia version, you must define a data container
inherit from <a class="reference external" href="https://ceandrade.github.io/brkga_mp_ipr_julia/guide/index.html#First-things-first:-basic-data-structures-and-decoder-function-1">AbstractInstance</a>,
and a decoder function. The reason you must do that is because structs in
Julia have no methods (but constructors), and the decoder function must take
both chromosome and input data in the call. In Python/C++, we can encapsulate the
input data into the decoder object, resulting in a much more clear API.</p>
<p>The basic form of a decoder should be:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">class</span> <span class="nc">Decoder</span><span class="p">():</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">pass</span>

    <span class="k">def</span> <span class="nf">decode</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chromosome</span><span class="p">:</span> <span class="n">BaseChromosome</span><span class="p">,</span> <span class="n">rewrite</span><span class="p">:</span> <span class="nb">bool</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
        <span class="k">return</span> <span class="mf">0.0</span>
</pre></div>
</div>
<p>The decoder <strong>must</strong> contain a <strong>decode()</strong> method that receives a
<code class="docutils literal notranslate"><span class="pre">BaseChromosome</span></code> reference and an <code class="docutils literal notranslate"><span class="pre">boolean</span></code>, and returns a float point
number. But before going further, let’s talk about the chromosome.</p>
<div class="section" id="the-chromosome-or-vector-of-doubles">
<h3>The chromosome or vector of doubles<a class="headerlink" href="#the-chromosome-or-vector-of-doubles" title="Permalink to this headline">¶</a></h3>
<p>Note that all long the BRKGA discussion, the chromosome is represented as a
vector of real numbers in the interval [0,1]. Indeed, we could use
straightforward <code class="docutils literal notranslate"><span class="pre">list</span></code>. However, sometimes is interesting to
keep more information inside the chromosome for further analysis, such as,
other solution metrics that not the main fitness value. For example, in a
scheduling problem, we may choose to keep both makespan and total completion
time metrics. Therefore, we chose to make the chromosome a “generic” data
structure in our design.</p>
<p>File <a class="reference external" href="https://github.com/ceandrade/brkga_mp_ipr_python/blob/master/brkga_mp_ipr/types.py">types.py</a> shows the
basic represetation of a chromosome:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">class</span> <span class="nc">BaseChromosome</span><span class="p">(</span><span class="nb">list</span><span class="p">):</span>
    <span class="k">pass</span>
</pre></div>
</div>
<p>Therefore, the <code class="docutils literal notranslate"><span class="pre">BaseChromosome</span></code> is a simple list, and can be tread as so in
your decoder. If this enough for you, you go already and use such a
definition. However, instead to redefine in your own code, <strong>we do recommend
to import and use the definition from</strong> <a class="reference external" href="https://github.com/ceandrade/brkga_mp_ipr_python/blob/master/brkga_mp_ipr/types.py">types.py</a>
since it is the same definition the main BRKGA-MP-IPR algorithm uses.</p>
<p>However, if you need more information to be tracked during the optimization,
you can redefine the chromosome. First, your definition must complain with
the <code class="docutils literal notranslate"><span class="pre">list</span></code> interface. The easiest way to do that is to inherit
from the <code class="docutils literal notranslate"><span class="pre">BaseChromosome</span></code>. For instance, assume we want to keep track of the
makespan and the total completion time for a scheduling problem. We can do
the following:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">class</span> <span class="nc">SchedulingChromosome</span><span class="p">(</span><span class="n">BaseChromosome</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">makespan</span> <span class="o">=</span> <span class="mf">0.0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">total_completion_time</span> <span class="o">=</span> <span class="mf">0.0</span>
</pre></div>
</div>
<p>Note that when subclassing BaseChromosome, we must define the method
<code class="docutils literal notranslate"><span class="pre">__init__(self,</span> <span class="pre">value)</span></code> and call the parent (<code class="docutils literal notranslate"><span class="pre">BaseChromosome</span></code>)
constructor. We need at least one argument to be passed to <code class="docutils literal notranslate"><span class="pre">BaseChromosome</span></code>
constructor. Note that the new custom chromosome type must be pass to the
main algorithm constructor (<code class="docutils literal notranslate"><span class="pre">BrkgaMpIpr.__init__</span></code>). Internally,
<code class="docutils literal notranslate"><span class="pre">BrkgaMpIpr</span></code> builds new chromosomes using
<code class="docutils literal notranslate"><span class="pre">CustomChromosomeType(&lt;list&gt;)</span></code>, where <code class="docutils literal notranslate"><span class="pre">&lt;list&gt;</span></code> is a list of floats in the
interval [0, 1].</p>
</div>
<div class="section" id="back-to-the-decoder">
<h3>Back to the decoder<a class="headerlink" href="#back-to-the-decoder" title="Permalink to this headline">¶</a></h3>
<p>Again, the decoder is the heart of a BRKGA. An easy way to keep the API clean
is to define a decoder that has a reference for the input data. This is a TSP
decoder defined on file <a class="reference external" href="https://github.com/ceandrade/brkga_mp_ipr_python/blob/master/examples/tsp/tsp_decoder.py">examples/tsp/tsp_decoder.py</a>:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">class</span> <span class="nc">TSPDecoder</span><span class="p">():</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">instance</span><span class="p">:</span> <span class="n">TSPInstance</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">instance</span> <span class="o">=</span> <span class="n">instance</span>

    <span class="k">def</span> <span class="nf">decode</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chromosome</span><span class="p">:</span> <span class="n">BaseChromosome</span><span class="p">,</span> <span class="n">rewrite</span><span class="p">:</span> <span class="nb">bool</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
        <span class="n">permutation</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
            <span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="n">index</span><span class="p">)</span> <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">key</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">chromosome</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="n">cost</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">instance</span><span class="o">.</span><span class="n">distance</span><span class="p">(</span><span class="n">permutation</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span> <span class="n">permutation</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="mi">1</span><span class="p">])</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">permutation</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">):</span>
            <span class="n">cost</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">instance</span><span class="o">.</span><span class="n">distance</span><span class="p">(</span><span class="n">permutation</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span>
                                           <span class="n">permutation</span><span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">][</span><span class="mi">1</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">cost</span>
</pre></div>
</div>
<p>Note that <code class="docutils literal notranslate"><span class="pre">TSPDecoder</span></code> get a reference to <code class="docutils literal notranslate"><span class="pre">TSPInstance</span></code> (from
<a class="reference external" href="https://github.com/ceandrade/brkga_mp_ipr_python/blob/master/examples/tsp/tsp_instance.py">examples/tsp/tsp_instance.py</a>),
that holds the input data. Therefore, <code class="docutils literal notranslate"><span class="pre">TSPDecoder</span></code> has direct access to the
data for optimization. This approach also benefits cache efficiency, mainly
when multiple threads are used for decoding, i.e., several threads can use
the same read-only data already in the cache, which speeds up the
optimization.</p>
<p>The decode method also has a <code class="docutils literal notranslate"><span class="pre">rewrite</span></code> argument that indicates if the decoder
should rewrite the chromosome, in case of local search / local improvements be
performed during the decoder process. This flag is critical if you intend to
use the Implicit Path Relink (not implemented in Python version yet).
Even though you do not rewrite the chromosome in your decoder, you must provide
such signature for API compatibility.</p>
<p>The decoder must return a <code class="docutils literal notranslate"><span class="pre">float</span></code> that is used as the <strong>fitness</strong> to rank
the chromosomes. In general, fitness is the cost/value of the solution, but you
may want to use it to penalize solutions that violate the problem constraints,
for example.</p>
<p>In our TSP example, we have a very simple decoder that generates a permutation
of nodes, and compute the cost of the cycle from that permutation
(note that we don’t use the flag <code class="docutils literal notranslate"><span class="pre">rewrite</span></code> in that example).</p>
<p>With the instance data and the decoder ready, we can build the BRKGA data
structures and perform the optimization.</p>
</div>
</div>
<div class="section" id="building-brkga-mp-ipr-algorithm-object">
<span id="doxid-guide-1guide-brkga-object"></span><h2>Building BRKGA-MP-IPR algorithm object<a class="headerlink" href="#building-brkga-mp-ipr-algorithm-object" title="Permalink to this headline">¶</a></h2>
<p><cite>BrkgaMpIpr</cite> is the main object that
implements all BRKGA-MP-IPR algorithms such as evolution, path relink, and
other auxiliary procedures.</p>
<p>The first step is to call the algorithm constructor that has the following
signature:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">decoder</span><span class="p">:</span> <span class="nb">object</span><span class="p">,</span> <span class="n">sense</span><span class="p">:</span> <span class="n">Sense</span><span class="p">,</span> <span class="n">seed</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
             <span class="n">chromosome_size</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">params</span><span class="p">:</span> <span class="n">BrkgaParams</span><span class="p">,</span>
             <span class="n">evolutionary_mechanism_on</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
             <span class="n">chrmosome_type</span><span class="p">:</span> <span class="nb">type</span> <span class="o">=</span> <span class="n">BaseChromosome</span><span class="p">)</span>
</pre></div>
</div>
<p>The first argument is the decoder object that must implement the <code class="docutils literal notranslate"><span class="pre">decode()</span></code>
method as discussed before. You also must indicate whether you are minimizing
or maximizing through parameter <code class="docutils literal notranslate"><span class="pre">BRKGA::Sense</span></code>.</p>
<p>A good seed also must be provided for the (pseudo) random number generator
(according to <a class="reference external" href="http://doi.acm.org/10.1145/1276927.1276928">this paper</a>).
BrkgaMpIpr uses the Mersenne Twister engine
<a class="reference external" href="http://dx.doi.org/10.1145/272991.272995">[1]</a>
<a class="reference external" href="https://en.wikipedia.org/wiki/Mersenne_Twister">[2]</a>
from the standard Python library
<a class="reference external" href="https://docs.python.org/3.7/library/random.html">[3]</a></p>
<p>The <code class="docutils literal notranslate"><span class="pre">chromosome_size</span></code> also must be given. It indicates the length of each
chromosome in the population. In general, this size depends on the instance
and how the decoder works. The constructor also takes a <code class="docutils literal notranslate"><span class="pre">BrkgaParams</span></code>
object that holds several parameters. We will take about that later.</p>
<p>Another common argument is <code class="docutils literal notranslate"><span class="pre">evolutionary_mechanism_on</span></code> which is enabled by
default. When disabled, no evolution is performed. The algorithm only decodes
the chromosomes and ranks them. On each generation, all population is replaced
excluding the best chromosome. This flag helps on implementations of simple
multi-start algorithms.</p>
<p>Finally, when using custom chromosomes, the user must provide the its
class/type. As explained before, internally, <code class="docutils literal notranslate"><span class="pre">BrkgaMpIpr</span></code> builds new
chromosomes using <code class="docutils literal notranslate"><span class="pre">CustomChromosomeType(&lt;list&gt;)</span></code>, where <code class="docutils literal notranslate"><span class="pre">&lt;list&gt;</span></code> is a
list of floats in the interval [0, 1].</p>
<p>All BRKGA and Path Relink hyper-parameters are stored in a <code class="docutils literal notranslate"><span class="pre">BrkgaParams</span></code>
object. Such objects can be read and write from plain text files or can be
created by hand by the user. There is also a companion
<code class="docutils literal notranslate"><span class="pre">ExternalControlParams</span></code> object that stores extra control parameters that
can be used outside the <code class="docutils literal notranslate"><span class="pre">BrkgaMpIpr</span></code> to control several aspects of the
optimization. For instance, interval to apply path relink, reset the
population, perform population migration, among others. This is how a
configuration file looks like (see <a class="reference external" href="https://github.com/ceandrade/brkga_mp_ipr_python/blob/master/examples/tsp/config.conf">config.conf</a>
for a commented version):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">population_size</span> <span class="mi">2000</span>
<span class="n">elite_percentage</span> <span class="mf">0.30</span>
<span class="n">mutants_percentage</span> <span class="mf">0.15</span>
<span class="n">num_elite_parents</span> <span class="mi">2</span>
<span class="n">total_parents</span> <span class="mi">3</span>
<span class="n">bias_type</span> <span class="n">LOGINVERSE</span>
<span class="n">num_independent_populations</span> <span class="mi">3</span>
<span class="n">pr_number_pairs</span> <span class="mi">0</span>
<span class="n">pr_minimum_distance</span> <span class="mf">0.15</span>
<span class="n">pr_type</span> <span class="n">PERMUTATION</span>
<span class="n">pr_selection</span> <span class="n">BESTSOLUTION</span>
<span class="n">alpha_block_size</span> <span class="mf">1.0</span>
<span class="n">pr_percentage</span> <span class="mf">1.0</span>
<span class="n">exchange_interval</span> <span class="mi">200</span>
<span class="n">num_exchange_indivuduals</span> <span class="mi">2</span>
<span class="n">reset_interval</span> <span class="mi">600</span>
</pre></div>
</div>
<p>To read this file, you can use the function <code class="docutils literal notranslate"><span class="pre">load_configuration()</span></code>, which
returns a tuple <code class="docutils literal notranslate"><span class="pre">(BrkgaParams,</span> <span class="pre">ExternalControlParams)</span></code>. When reading such
file, the function ignores all blank lines, and lines starting with <code class="docutils literal notranslate"><span class="pre">#</span></code>. As
commented before, <code class="docutils literal notranslate"><span class="pre">BrkgaParams</span></code> contains all hyper-parameters regarding
BRKGA and IPR methods and <code class="docutils literal notranslate"><span class="pre">ExternalControlParams</span></code> contains extra control
parameters, and although their presence is required on the config file, they
are not mandatory to the BRKGA-MP-IPR itself.</p>
<p>Let’s take a look in the example from <a class="reference external" href="https://github.com/ceandrade/brkga_mp_ipr_python/blob/master/examples/tsp/main_minimal.py">main_minimal.py</a>:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">seed</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="n">configuration_file</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
<span class="n">num_generations</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
<span class="n">instance_file</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span>

<span class="n">instance</span> <span class="o">=</span> <span class="n">TSPInstance</span><span class="p">(</span><span class="n">instance_file</span><span class="p">)</span>

<span class="n">brkga_params</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">load_configuration</span><span class="p">(</span><span class="n">configuration_file</span><span class="p">)</span>

<span class="n">decoder</span> <span class="o">=</span> <span class="n">TSPDecoder</span><span class="p">(</span><span class="n">instance</span><span class="p">)</span>

<span class="n">brkga</span> <span class="o">=</span> <span class="n">BrkgaMpIpr</span><span class="p">(</span>
    <span class="n">decoder</span><span class="o">=</span><span class="n">decoder</span><span class="p">,</span>
    <span class="n">sense</span><span class="o">=</span><span class="n">Sense</span><span class="o">.</span><span class="n">MINIMIZE</span><span class="p">,</span>
    <span class="n">seed</span><span class="o">=</span><span class="n">seed</span><span class="p">,</span>
    <span class="n">chromosome_size</span><span class="o">=</span><span class="n">instance</span><span class="o">.</span><span class="n">num_nodes</span><span class="p">,</span>
    <span class="n">params</span><span class="o">=</span><span class="n">brkga_params</span>
<span class="p">)</span>
</pre></div>
</div>
<p>This code gets some arguments from the command line and loads a TSP instance.
After that, it reads the BRKGA parameters from the configuration file. Since in
this example, we only care about the BRKGA parameters, we ignore the control
parameters. We then build the decoder object, and the BRKGA
algorithm. Since we are looking for cycles of minimum cost, we ask for the
algorithm <code class="docutils literal notranslate"><span class="pre">MINIMIZE</span></code>. The starting seed is also given. Since <code class="docutils literal notranslate"><span class="pre">TSPDecode</span></code>
considers each chromosome key as a node/city, the length of the chromosome must
be the number of nodes, i.e., <code class="docutils literal notranslate"><span class="pre">instance.num_nodes</span></code>. Finally, we also pass the
BRKGA parameters.</p>
<p>Now, we have a <code class="docutils literal notranslate"><span class="pre">BrkgaMpIpr</span></code> which will be used to call all other functions
during the optimization. Note that we can build several <code class="docutils literal notranslate"><span class="pre">BrkgaMpIpr</span></code>
objects using different parameters, decoders, or instance data. These
structures can be evolved in parallel and mixed-and-matched at your will.
Each one holds a self-contained BRKGA state including populations, fitness
information, and a state of the random number generator.</p>
</div>
<div class="section" id="initialization-and-warm-start-solutions">
<span id="doxid-guide-1guide-algo-init"></span><h2>Initialization and Warm-start solutions<a class="headerlink" href="#initialization-and-warm-start-solutions" title="Permalink to this headline">¶</a></h2>
<p>Before starting the optimization, we need to initialize the <code class="docutils literal notranslate"><span class="pre">BrkgaMpIpr</span></code>
algorithm state using <code class="docutils literal notranslate"><span class="pre">BrkgaMpIpr.initialize()</span></code> method. This procedure
initializes the populations and others data structures of the BRKGA. If an
initial population (warm start) is supplied, the initialization method
completes the remaining individuals, if they do not exist. This method also
performs the initial decoding of the chromosomes. Therefore, depending on the
decoder implementation, this can take a while, and you may want to time such
procedure. Assuming that <code class="docutils literal notranslate"><span class="pre">brkga</span></code> is our <code class="docutils literal notranslate"><span class="pre">BrkgaMpIpr</span></code> object, the syntax
is pretty straightforward:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">brkga</span><span class="o">.</span><span class="n">initialize</span><span class="p">()</span>
</pre></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p><code class="docutils literal notranslate"><span class="pre">initialize()</span></code> must be called before any optimization methods.</p>
</div>
<div class="section" id="warm-start-solutions">
<h3>Warm-start solutions<a class="headerlink" href="#warm-start-solutions" title="Permalink to this headline">¶</a></h3>
<p>One good strategy is to bootstrap the main optimization algorithm with good
solutions from fast heuristics
[<a class="reference external" href="http://dx.doi.org/10.1002/net.21685">1</a>,
<a class="reference external" href="http://dx.doi.org/10.1016/j.ejor.2017.10.045">2</a>,
<a class="reference external" href="http://dx.doi.org/10.1016/j.ejor.2017.10.045">3</a>]
or even from relaxations of integer linear programming models
<a class="reference external" href="http://dx.doi.org/10.1162/EVCO_a_00138">[4]</a>.</p>
<p>To do it, you must set these initial solutions before call <code class="docutils literal notranslate"><span class="pre">initialize()</span></code>.
Since BrkgaMpIpr does not know the problem structure, you must <em>encode</em> the
warm-start solution as chromosomes (vectors in the interval [0, 1]). In other
words, you must do the inverse process that your decoder does. For instance,
this is a piece of code from <a class="reference external" href="https://github.com/ceandrade/brkga_mp_ipr_python/blob/master/examples/tsp/main_complete.py">main_complete.py</a>
showing this process:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">initial_cost</span><span class="p">,</span> <span class="n">initial_tour</span> <span class="o">=</span> <span class="n">greedy_tour</span><span class="p">(</span><span class="n">instance</span><span class="p">)</span>
<span class="o">...</span>

<span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
<span class="n">keys</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">()</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">instance</span><span class="o">.</span><span class="n">num_nodes</span><span class="p">)])</span>

<span class="n">initial_chromosome</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">instance</span><span class="o">.</span><span class="n">num_nodes</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">instance</span><span class="o">.</span><span class="n">num_nodes</span><span class="p">):</span>
    <span class="n">initial_chromosome</span><span class="p">[</span><span class="n">initial_tour</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span> <span class="o">=</span> <span class="n">keys</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>

<span class="n">brkga</span><span class="o">.</span><span class="n">set_initial_population</span><span class="p">([</span><span class="n">initial_chromosome</span><span class="p">])</span>
</pre></div>
</div>
<p>Here, we create one incumbent solution using the greedy heuristic
<code class="docutils literal notranslate"><span class="pre">greedy_tour()</span></code> <a class="reference external" href="https://github.com/ceandrade/brkga_mp_ipr_python/blob/master/examples/tsp/greedy_tour.py">found here</a>.
It gives us the <code class="docutils literal notranslate"><span class="pre">initial_cost</span></code> of the tour represented by the sequence of
nodes <code class="docutils literal notranslate"><span class="pre">initial_tour</span></code>. In the next lines, we encode <code class="docutils literal notranslate"><span class="pre">initial_tour</span></code>. First,
we create a vector of sorted random <code class="docutils literal notranslate"><span class="pre">keys</span></code> using a local random number
generator. For that, we create the vector <code class="docutils literal notranslate"><span class="pre">keys</span></code>, and fill up <code class="docutils literal notranslate"><span class="pre">keys</span></code> with
random numbers in the interval [0,1]. Once we have the keys, we sort them as
<code class="docutils literal notranslate"><span class="pre">TSPDecoder.decode()</span></code> does. We then create the <code class="docutils literal notranslate"><span class="pre">initial_chromosome</span></code>, and
fill it up with <code class="docutils literal notranslate"><span class="pre">keys</span></code> according to the nodes’ order in <code class="docutils literal notranslate"><span class="pre">initial_tour</span></code>.
Finally, we call `` BrkgaMpIpr.set_initial_population()`` to assign the
incumbent to the initial population. Note that we enclose the initial
solution inside a vector of chromosomes, since <code class="docutils literal notranslate"><span class="pre">set_initial_population()</span></code>
may take more than one starting solution. See its signature:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">set_initial_population</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chromosomes</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">BaseChromosome</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span>
</pre></div>
</div>
<p>Indeed, you can have as much warm-start solutions as you like, limited to the
size of the population. Just remember:</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p><code class="docutils literal notranslate"><span class="pre">set_initial_population()</span></code> must be called <strong>BEFORE</strong> <code class="docutils literal notranslate"><span class="pre">initialize()</span></code>.</p>
</div>
</div>
</div>
<div class="section" id="optimization-time-evolving-the-population">
<span id="doxid-guide-1guide-opt"></span><h2>Optimization time: evolving the population<a class="headerlink" href="#optimization-time-evolving-the-population" title="Permalink to this headline">¶</a></h2>
<p>Once all data is set up, it is time to evolve the population and perform other
operations like path-relinking, shaking, migration, and others. The call is
pretty simple:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">brkga</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span><span class="n">num_generations</span><span class="p">);</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">BrkgaMpIpr.evolve()</span></code>
evolves all populations for <code class="docutils literal notranslate"><span class="pre">num_generations</span></code>. If <code class="docutils literal notranslate"><span class="pre">num_genertions</span></code> is
omitted, <code class="docutils literal notranslate"><span class="pre">evolve()</span></code> evolves only one generation.</p>
<p>For example, in <a class="reference external" href="https://github.com/ceandrade/brkga_mp_ipr_python/blob/master/examples/tsp/main_mininal.py">main_minimal.py</a>,
we just evolve the population for a given number of generations directly and
then extract the best solution cost.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">brkga</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span><span class="n">num_generations</span><span class="p">);</span>
<span class="n">best_cost</span> <span class="o">=</span> <span class="n">brkga</span><span class="o">.</span><span class="n">get_best_fitness</span><span class="p">()</span>
</pre></div>
</div>
<p>On <a class="reference external" href="https://github.com/ceandrade/brkga_mp_ipr_python/blob/master/examples/tsp/main_complete.py">main_complete.py</a>
we have fine-grained control on the optimization. There, we have a main loop
that evolves the population one generation at a time and performs several
operations as to hold the best solution, to check whether it is time for path
relink, population reset, among others. The advantage of that code is that we
can track all optimization details, and I do recommend similar style for
experimentation.</p>
</div>
<div class="section" id="accessing-solutions-chromosomes">
<span id="doxid-guide-1guide-access-solutions"></span><h2>Accessing solutions/chromosomes<a class="headerlink" href="#accessing-solutions-chromosomes" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">BrkgaMpIpr</span></code> offers several mechanisms to access a variaty of data during
the optimization. Most common, we want to access the best chromosome after some
iterations. You can use the companion functions:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">get_best_fitness</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span>

<span class="k">def</span> <span class="nf">get_best_chromosome</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">BaseChromosome</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">get_best_fitness()</span></code>
returns the value/fitness of the best chromosome across all populations.</p>
<p><code class="docutils literal notranslate"><span class="pre">get_best_chromosome()</span></code> returns a <em>deep copy</em> of the best chromosome across
all populations. You may want to extract an actual solution from such
chromosome, i.e., to apply a decoding function that returns the actual
solution instead only its value.</p>
<p>You may also want to get a reference of specific chromosome for a given
population using <code class="docutils literal notranslate"><span class="pre">BrkgaMpIpr.get_chromosome()</span></code>.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">get_chromosome</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">population_index</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">position</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">BaseChromosome</span>
</pre></div>
</div>
<p>For example, you can get the 3rd best chromosome from the 2nd population using</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">third_best</span> <span class="o">=</span> <span class="n">brkga</span><span class="o">.</span><span class="n">get_chromosome</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Just remember that Python is zero-indexed. So, the first population index is 0
(zero), the second population index is 1 (one), and so forth. The same happens
for the chromosomes.</p>
</div>
<p>Now, suppose you get such chromosome or chromosomes and apply a quick local
search procedure on them. It may be useful to reinsert such new solutions in
the BRKGA population for the next
evolutionary cycles. You can do that using
<code class="docutils literal notranslate"><span class="pre">BrkgaMpIpr.inject_chromosome()</span></code>:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">inject_chromosome</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chromosome</span><span class="p">:</span> <span class="n">BaseChromosome</span><span class="p">,</span>
                      <span class="n">population_index</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">position</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
                      <span class="n">fitness</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">inf</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span>
</pre></div>
</div>
<p>Note that the chromosome is put in a specific position of a given population.
If you do not provide the fitness, <code class="docutils literal notranslate"><span class="pre">inject_chromosome()</span></code> will decode the
injected chromosome. For example, assuming the <code class="docutils literal notranslate"><span class="pre">brkga</span></code> is your
BRKGA-MP-IPR object and <code class="docutils literal notranslate"><span class="pre">brkga_params</span></code> is your <code class="docutils literal notranslate"><span class="pre">BrkgaParams</span></code> object, the
following code injects the random chromosome <code class="docutils literal notranslate"><span class="pre">keys</span></code> into the population #1 in
the last position (<code class="docutils literal notranslate"><span class="pre">population_size</span></code>), i.e., it will replace the worst
solution by a random one:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
<span class="n">keys</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">()</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">instance</span><span class="o">.</span><span class="n">num_nodes</span><span class="p">)])</span>
<span class="n">brkga</span><span class="o">.</span><span class="n">inject_chromosome</span><span class="p">(</span><span class="n">keys</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">brkga_params</span><span class="o">.</span><span class="n">population_size</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="implicit-path-relink">
<span id="doxid-guide-1guide-ipr"></span><h2>Implicit Path Relink<a class="headerlink" href="#implicit-path-relink" title="Permalink to this headline">¶</a></h2>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The Implicit Path Relink is not implemented in the Python version yet.</p>
</div>
</div>
<div class="section" id="shaking-and-resetting">
<span id="doxid-guide-1guide-shaking-reset"></span><h2>Shaking and Resetting<a class="headerlink" href="#shaking-and-resetting" title="Permalink to this headline">¶</a></h2>
<p>Sometimes, BRKGA gets stuck, converging to local maxima/minima, for several
iterations. When such a situation happens, it is a good idea to perturb the
population, or even restart from a new one completely new. BrkgaMpIpr offers
<code class="docutils literal notranslate"><span class="pre">shake()</span></code> method, an improved variation of the original version
proposed in [this paper](<a class="reference external" href="http://dx.doi.org/10.1016/j.eswa.2019.03.007">http://dx.doi.org/10.1016/j.eswa.2019.03.007</a>).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Shaking is not implemented in the Python version yet.</p>
</div>
<p>Sometimes, even shaking the populations does not help to escape from local
maxima/minima. So, we need a drastic measure, restarting from scratch the role
population. This can be easily accomplished with
<code class="docutils literal notranslate"><span class="pre">BrkgaMpIpr.reset()</span></code>:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">brkga</span><span class="o">.</span><span class="n">reset</span><span class="p">()</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>When using <code class="docutils literal notranslate"><span class="pre">reset()</span></code>, all warm-start solutions provided by
<code class="docutils literal notranslate"><span class="pre">set_initial_population()</span></code> are discarded. You may use
<code class="docutils literal notranslate"><span class="pre">inject_chromosome()</span></code> to insert those solutions again.</p>
</div>
</div>
<div class="section" id="multi-population-and-migration">
<span id="doxid-guide-1guide-migration"></span><h2>Multi-population and migration<a class="headerlink" href="#multi-population-and-migration" title="Permalink to this headline">¶</a></h2>
<p>Multi-population or <em>island model</em> was introduced in genetic algorithms in
<a class="reference external" href="http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.36.7225">this paper</a>. The idea
is to evolve parallel and independent populations and, once a while, exchange
individuals among these populations. In several scenarios, this approach is
very beneficial for optimization.</p>
<p>BRKGA-MP-IPR is implemented using such island idea from the core. If you read
the guide until here, you may notice that several methods take into account
multiple populations. To use multiple populations, you must set
<code class="docutils literal notranslate"><span class="pre">BrkgaParams.num_independent_populations</span></code> with 2 ou more populations, and
build the BRKGA algorithm from such parameters.</p>
</div>
<div class="section" id="simulating-the-standard-brkga">
<span id="doxid-guide-1guide-standard-brkga"></span><h2>Simulating the standard BRKGA<a class="headerlink" href="#simulating-the-standard-brkga" title="Permalink to this headline">¶</a></h2>
<p>Sometimes, it is a good idea to test how the standard BRKGA algorithm performs
for a problem. You can use BrkgaMpIpr framework to quickly implement and test
a standard BRKGA.</p>
<p>First, you must guarantee that, during the crossover, the algorithm chooses
only one elite individual and only one non-elite individual. This is easily
accomplished setting <code class="docutils literal notranslate"><span class="pre">num_elite_parents</span> <span class="pre">=</span> <span class="pre">1</span></code> and <code class="docutils literal notranslate"><span class="pre">total_parents</span> <span class="pre">=</span> <span class="pre">2</span></code>. Then,
you must set up a bias function that ranks the elite and no-elite individual
according to the original BRKGA bias parameter <span class="math notranslate nohighlight">\(\rho\)</span> (rho).</p>
<p>You can use <code class="docutils literal notranslate"><span class="pre">BrkgaMpIpr.set_bias_custom_function()</span></code>
for that task. The given function receives the index of the chromosome and
returns a ranking for it. Such ranking is used in the roulette method to choose
the individual from which each allele comes to build the new chromosome. Since
we have one two individuals for crossover in the standard BRKGA, the bias
function must return the probability to one or other individual. In the
following code, we do that with a simple <code class="docutils literal notranslate"><span class="pre">if...else</span></code> lambda function.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create brkga_params by hand or reading from a file,</span>
<span class="c1"># then set the following by hand.</span>
<span class="n">brkga_params</span><span class="o">.</span><span class="n">num_elite_parents</span> <span class="o">=</span> <span class="mi">1</span>
<span class="n">brkga_params</span><span class="o">.</span><span class="n">total_parents</span> <span class="o">=</span> <span class="mi">2</span>

<span class="n">rho</span> <span class="o">=</span> <span class="mf">0.75</span><span class="p">;</span>
<span class="n">brkga</span><span class="o">.</span><span class="n">set_bias_custom_function</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">rho</span> <span class="k">if</span> <span class="n">x</span> <span class="o">==</span> <span class="mi">1</span> <span class="k">else</span> <span class="mf">1.0</span> <span class="o">-</span> <span class="n">rho</span><span class="p">)</span>
<span class="n">brkga</span><span class="o">.</span><span class="n">initialize</span><span class="p">()</span>
</pre></div>
</div>
<p>Here, we first set the <code class="docutils literal notranslate"><span class="pre">num_elite_parents</span> <span class="pre">=</span> <span class="pre">1</span></code> and <code class="docutils literal notranslate"><span class="pre">total_parents</span> <span class="pre">=</span> <span class="pre">2</span></code> as
explained before. Following, we set a variable <code class="docutils literal notranslate"><span class="pre">rho</span> <span class="pre">=</span> <span class="pre">0.75</span></code>. This is the
<span class="math notranslate nohighlight">\(\rho\)</span> from standard BRKGA, and you may set it as you wish. Then, we set
the bias function as a very simple lambda function:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">rho</span> <span class="k">if</span> <span class="n">x</span> <span class="o">==</span> <span class="mi">1</span> <span class="k">else</span> <span class="mf">1.0</span> <span class="o">-</span> <span class="n">rho</span>
</pre></div>
</div>
<p>So, if the index of the chromosome is 1 (elite individual), it gets a 0.75
rank/probability. If the index is 2 (non-elite individual), the chromosome gets
0.25 rank/probability.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>All these operations must be done before calling <code class="docutils literal notranslate"><span class="pre">initialize()</span></code>.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Note that we consider the index 1 as the elite individual instead index
0, and index 2 to the non-elite individual opposed to index 1. The reason
for this is that, internally, BRKGA always pass <code class="docutils literal notranslate"><span class="pre">r</span> <span class="pre">&gt;</span> <span class="pre">0</span></code> to the bias
function to avoid division-by-zero exceptions.</p>
</div>
</div>
<div class="section" id="reading-and-writing-parameters">
<span id="doxid-guide-1guide-parameters"></span><h2>Reading and writing parameters<a class="headerlink" href="#reading-and-writing-parameters" title="Permalink to this headline">¶</a></h2>
<p>Although we can build the BRKGA algorithm data by set up a <code class="docutils literal notranslate"><span class="pre">BrkgaParams</span></code>
object manually, the easiest way to do so is to read such parameters from a
configuration file. For this, we can use <code class="docutils literal notranslate"><span class="pre">read_configuration()</span></code> that reads
a simple plain text file and returns a tuple of <code class="docutils literal notranslate"><span class="pre">BrkgaParams</span></code> and
<code class="docutils literal notranslate"><span class="pre">ExternalControlParams</span></code> objects. For instance,</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">brkga_params</span><span class="p">,</span> <span class="n">control_params</span> <span class="o">=</span> <span class="n">read_configuration</span> <span class="p">(</span><span class="s2">&quot;tuned_conf.txt&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>The configuration file must be plain text such that contains pairs of
parameter name and value. This file must list all fields from <code class="docutils literal notranslate"><span class="pre">BrkgaParams</span></code>
<code class="docutils literal notranslate"><span class="pre">ExternalControlParams</span></code>, even though you do not use each one at this
moment. In <a class="reference external" href="https://github.com/ceandrade/brkga_mp_ipr_python/tree/master/examples/tsp">examples folder</a>,
we have <a class="reference external" href="https://github.com/ceandrade/brkga_mp_ipr_python/blob/master/examples/tsp/config.conf">config.conf</a>
that looks like this:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">population_size</span> <span class="mi">2000</span>
<span class="n">elite_percentage</span> <span class="mf">0.30</span>
<span class="n">mutants_percentage</span> <span class="mf">0.15</span>
<span class="n">num_elite_parents</span> <span class="mi">2</span>
<span class="n">total_parents</span> <span class="mi">3</span>
<span class="n">bias_type</span> <span class="n">LOGINVERSE</span>
<span class="n">num_independent_populations</span> <span class="mi">3</span>
<span class="n">pr_number_pairs</span> <span class="mi">0</span>
<span class="n">pr_minimum_distance</span> <span class="mf">0.15</span>
<span class="n">pr_type</span> <span class="n">PERMUTATION</span>
<span class="n">pr_selection</span> <span class="n">BESTSOLUTION</span>
<span class="n">alpha_block_size</span> <span class="mf">1.0</span>
<span class="n">pr_percentage</span> <span class="mf">1.0</span>
<span class="n">exchange_interval</span> <span class="mi">200</span>
<span class="n">num_exchange_indivuduals</span> <span class="mi">2</span>
<span class="n">reset_interval</span> <span class="mi">600</span>
</pre></div>
</div>
<p>It does not matter whether we use lower or upper cases. Blank lines and lines
starting with <code class="docutils literal notranslate"><span class="pre">#</span></code> are ignored. The order of the parameters should not matter
either. And, finally, this file should be readble for both C++, Julia,
and Python framework versions.</p>
<p>In some cases, you define some of the parameters at the running time, and you
may want to save them for debug or posterior use. To do so, you can use
<code class="docutils literal notranslate"><span class="pre">write_configuration()``call</span> <span class="pre">upon</span> <span class="pre">a</span> <span class="pre">``BrkgaParams</span></code> and
<code class="docutils literal notranslate"><span class="pre">ExternalControlParams</span></code> objects.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">write_configuration</span><span class="p">(</span><span class="s2">&quot;test.conf&quot;</span><span class="p">,</span> <span class="n">brkga_params</span><span class="p">,</span> <span class="n">control_params</span><span class="p">)</span>
<span class="c1"># or</span>
<span class="n">write_configuration</span><span class="p">(</span><span class="s2">&quot;test.conf&quot;</span><span class="p">,</span> <span class="n">brkga_params</span><span class="p">,</span> <span class="n">ExternalControlParams</span><span class="p">())</span>
</pre></div>
</div>
<p>In the last line, default values are used for <code class="docutils literal notranslate"><span class="pre">ExternalControlParams</span></code>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><code class="docutils literal notranslate"><span class="pre">write_configuration()</span></code> rewrites the given file. So, watch out to not lose
previous configurations.</p>
</div>
</div>
<div class="section" id="probable-valuable-tips">
<span id="doxid-guide-1guide-tips"></span><h2>(Probable Valuable) Tips<a class="headerlink" href="#probable-valuable-tips" title="Permalink to this headline">¶</a></h2>
<div class="section" id="algorithm-warmup">
<h3>Algorithm warmup<a class="headerlink" href="#algorithm-warmup" title="Permalink to this headline">¶</a></h3>
<p>While in Julia framework version is primordial to do a dry-run to precompile
all functions, in C++ and Python this warmup is not necessary. However,
Python compiles all scripts not previously compiled, and this can take some
time. Therefore, it is a good idea to warm up your code. Another advantage is
the memory location effects of our data (principle of locality), that can be
brought closer to the processor (L2/L3 caches) during the running.
Obliviously, this depends on how you implement and use your data structures.</p>
<p>In <a class="reference external" href="https://github.com/ceandrade/brkga_mp_ipr_python/blob/master/examples/tsp/main_complete.py">main_complete.py</a>,
we have the following piece of code to warmup mainly the decoder and other
functions. Note that we just deep-copy <code class="docutils literal notranslate"><span class="pre">brkga</span></code>, and then, we may lose
the principle of locality.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">bogus_alg</span> <span class="o">=</span> <span class="n">deepcopy</span><span class="p">(</span><span class="n">brkga</span><span class="p">)</span>
<span class="n">bogus_alg</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
<span class="c1"># TODO (ceandrade): warm up path relink functions.</span>
<span class="c1"># bogus_alg.path_relink(brkga_params.pr_type, brkga_params.pr_selection,</span>
<span class="c1">#              (x, y) -&gt; 1.0, (x, y) -&gt; true, 0, 0.5, 1, 10.0, 1.0)</span>
<span class="n">bogus_alg</span><span class="o">.</span><span class="n">get_best_fitness</span><span class="p">()</span>
<span class="n">bogus_alg</span><span class="o">.</span><span class="n">get_best_chromosome</span><span class="p">()</span>
<span class="n">bogus_alg</span> <span class="o">=</span> <span class="kc">None</span>
</pre></div>
</div>
</div>
<div class="section" id="complex-decoders-and-timing">
<h3>Complex decoders and timing<a class="headerlink" href="#complex-decoders-and-timing" title="Permalink to this headline">¶</a></h3>
<p>Some problems require complex decoders while for others, the decoder contains
local search procedures, that can be time-consuming. In general, the decoding
is the most time-expensive component of a BRKGA algorithm, and it may skew some
stopping criteria based on running time. Therefore, if your decoder is
time-consuming, it is a good idea to implement a timer or chronometer kind of
thing inside the decoder.</p>
<p>Testing for stopping time uses several CPU cycles, and you need to be careful
when/where to test it, otherwise, you spend all the optimization time doing
system calls to the clock.</p>
<p>IMHO, the most effective way to do it is to test time at the very end of the
decoding. If the current time is larger than the maximum time allowed, simple
return <code class="docutils literal notranslate"><span class="pre">Inf</span></code> or <code class="docutils literal notranslate"><span class="pre">-Inf</span></code> according to your optimization direction. In this
way, we make the solution <strong>invalid</strong> since it violates the maximum time
allowed. The BRKGA framework takes care of the rest.</p>
</div>
<div class="section" id="multi-threading">
<h3>Multi-threading<a class="headerlink" href="#multi-threading" title="Permalink to this headline">¶</a></h3>
<p>BRKGA multi-threading decoding is not implemented in Python at this time.</p>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="page_license.html" class="btn btn-neutral float-right" title="BRKGA-MP-IPR License" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="index.html" class="btn btn-neutral float-left" title="BRKGA-MP-IPR Guide and Documentation - Python version" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Carlos E. Andrade

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>